expression/constant_propagation.go

```diff
@@ -290,37 +308,276 @@ func (s *propagateConstantSolver) solve(conditions []Expression) []Expression {
+	retMapper := make(map[int]*Constant)
+	retMapper = s.pickNewEQCondsFunc(retMapper, visited, true)
+	if retMapper == nil {
+		// Filter is constant false, enforce early termination.
+		return nil
+	}
+	retMapper = s.pickNewEQCondsFunc(retMapper, visited, false)
+	return retMapper
 }
 
-func (s *propagateConstantSolver) insertCol(col *Column) {
-	code := col.HashCode(nil)
-	_, ok := s.colMapper[string(code)]
-	if !ok {
-		s.colMapper[string(code)] = len(s.colMapper)
-		s.columns = append(s.columns, col)
+// propagateConstantEQ propagates expressions like `outerCol = const` by substituting `outerCol` in *JOIN* condition
+// with `const`, the procedure repeats multiple times.
+func (s *propOuterJoinConstSolver) propagateConstantEQ() {
+	s.eqList = make([]*Constant, len(s.columns))
+	lenFilters := len(s.fConds)
+	visited := make([]bool, lenFilters+len(s.jConds))
+	for i := 0; i < MaxPropagateColsCnt; i++ {
+		mapper := s.pickNewEQConds(visited)
+		if len(mapper) == 0 {
+			return
+		}
+		cols := make([]*Column, 0, len(mapper))
+		cons := make([]Expression, 0, len(mapper))
+		for id, con := range mapper {
+			cols = append(cols, s.columns[id])
+			cons = append(cons, con)
+		}
+		for i, cond := range s.jConds {
+			if !visited[i+lenFilters] {
+				s.jConds[i] = ColumnSubstitute(cond, NewSchema(cols...), cons)
+			}
+		}
 	}
 }
 
-// PropagateConstant propagate constant values of deterministic predicates in a condition.
-func PropagateConstant(ctx sessionctx.Context, conditions []Expression) []Expression {
-	solver := &propagateConstantSolver{
-		colMapper: make(map[string]int),
-		ctx:       ctx,
+func (s *propOuterJoinConstSolver) colsFromOuterAndInner(col1, col2 *Column) (*Column, *Column) {
+	if s.outerSchema.Contains(col1) && s.innerSchema.Contains(col2) {
+		return col1, col2
 	}
-	return solver.solve(conditions)
+	if s.outerSchema.Contains(col2) && s.innerSchema.Contains(col1) {
+		return col2, col1
+	}
+	return nil, nil
+}
+
+// validColEqualCond checks if expression is column equql condition that we can use for constant
+// propagation over outer join. We only use expression like `outerCol = innerCol`, for expressions like
+// `outerCol1 = outerCol2` or `innerCol1 = innerCol2`, they do not help deriving new inner table conditions
+// which can be pushed down to children plan nodes, so we do not pick them.
+func (s *propOuterJoinConstSolver) validColEqualCond(cond Expression) (*Column, *Column) {
+	if fun, ok := cond.(*ScalarFunction); ok && fun.FuncName.L == ast.EQ {
+		lCol, lOk := fun.GetArgs()[0].(*Column)
+		rCol, rOk := fun.GetArgs()[1].(*Column)
+		if lOk && rOk {
+			return s.colsFromOuterAndInner(lCol, rCol)
+		}
+	}
+	return nil, nil
+
+}
+
+// deriveConds given `outerCol = innerCol`, derive new expression for specified conditions.
+func (s *propOuterJoinConstSolver) deriveConds(outerCol, innerCol *Column, schema *Schema, fCondsOffset int, visited []bool, fConds bool) []bool {
```

Comment:
>visited = s.deriveConds(outerCol, innerCol, mergedSchema, lenJoinConds, visited, false)

Here `lenJoinConds` is passed. Why here named `fCondsOffset`?

Reply:
Hmm, because it is indeed an offset inside `deriveConds` definition, while it is computed from `lenJoinConds` in caller...